### PR TITLE
metadata-service: validate that the dockerImageTag is not decremented

### LINKED ISSUE #42529: Metadata validation: make sure root dockerImageTag is never decremented
Flaw:
The source of truth to determine which connector version is about to get released is the value of the root dockerImageTag field. If this field is used for pinning a connector to a previous version it will first be used in attempting a publish. This can be dangerous if the pinned version is not released: it will be published on merge with the current version. 

Action item:

We should prevent developers from incorrectly pinning a connector via CI. We could implement a new metadata validation to make sure that the root dockerImageTag is never decremented

### PR DESCRIPTION
**Pull request by alafanechere**
## What
Closes https://github.com/airbytehq/airbyte/issues/42529

We want to prevent developers from decrementing the `dockerImageTag` value. This is to avoid version pinning errors.

## How
* Fetch the latest released image tag from DockerHub API
* Check if the current version is higher or equal to the latest released version: fail otherwise.

### CODE REVIEW

**Comment by alafanechere** [comment]
* **#42878** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/42878?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @alafanechere and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: airbyte-ci/connectors/metadata_service/lib/metadata_service/validators/metadata_validator.py [hunk 1]
```diff
@@ -175,13 +175,36 @@ def validate_pypi_only_for_python(
...
+    latest_released_version = get_latest_version_on_dockerhub(docker_image_name)
+    # This is happening when the connector has never been released to DockerHub
+    if not latest_released_version:
+        return True, None
+    if docker_image_tag == latest_released_version:
+        return True, None
+    current_semver_version = semver.Version.parse(docker_image_tag)
+    latest_released_semver_version = semver.Version.parse(latest_released_version)
+    if current_semver_version < latest_released_semver_version:
+        return False, f"The dockerImageTag value can't be decremented: it should be equal to or above {latest_released_version}."
```
Comment (clnoll) [inline airbyte-ci/connectors/metadata_service/lib/metadata_service/validators/metadata_validator.py:196]:
```suggestion
        return False, f"The dockerImageTag value can't be decremented: it should be equal to or above {latest_released_version} (got {current_semver_version})."
```

File: airbyte-ci/connectors/metadata_service/lib/metadata_service/docker_hub.py [hunk 2]
```diff
@@ -27,6 +27,75 @@ def get_docker_hub_auth_token() -> str:
...
+                get_docker_hub_tags_and_digests(
+                    image_name, retries=retries, wait_sec=wait_sec, next_page_url=next_page_url, tags_and_digests=tags_and_digests
+                )
+            )
+    return tags_and_digests
+
+
+def get_latest_version_on_dockerhub(image_name: str) -> str | None:
+    tags_and_digests = get_docker_hub_tags_and_digests(image_name, retries=3, wait_sec=30)
+    if latest_digest := tags_and_digests.get("latest"):
```
Comment (clnoll) [inline airbyte-ci/connectors/metadata_service/lib/metadata_service/docker_hub.py:92]: Not all of our docker repos have a `"latest"` tag, and we don't really rely on `"latest"` for getting the prod version of a connector (which is a good thing). So I feel like we shouldn't rely on it here, either. Can we check for the max version instead?

File: airbyte-ci/connectors/metadata_service/lib/metadata_service/docker_hub.py [hunk 3]
```diff
@@ -27,6 +27,75 @@ def get_docker_hub_auth_token() -> str:
...
+        tags_and_digest (Dict[str, str] | None, optional): The accumulated tags and digests for recursion. Defaults to None.
+
+    Returns:
+        Dict[str, str]: Mapping of image tag to digest
+    """
+    headers = get_docker_hub_headers()
+    tags_and_digests = tags_and_digests or {}
+
+    if not next_page_url:
+        tags_url = f"https://registry.hub.docker.com/v2/repositories/{image_name}/tags"
```
Comment (clnoll) [inline airbyte-ci/connectors/metadata_service/lib/metadata_service/docker_hub.py:64]: Can you link to the documentation for this?

I'm actually seeing a slightly different url [here](https://docs.docker.com/docker-hub/api/latest/#tag/repositories/paths/~1v2~1namespaces~1%7Bnamespace%7D~1repositories~1%7Brepository%7D~1tags/get), not sure what accounts for the difference.

File: airbyte-ci/connectors/metadata_service/lib/metadata_service/docker_hub.py [hunk 4]
```diff
@@ -27,6 +27,75 @@ def get_docker_hub_auth_token() -> str:
...
+        tags_url = next_page_url
+
+    # Allow for retries as the DockerHub API is not always reliable with returning the latest publish.
+    for _ in range(retries + 1):
+        response = requests.get(tags_url, headers=headers)
+        if response.ok:
+            break
+            # This is to handle the case when a connector has not ever been released yet.
+        if response.status_code == 404:
+            return tags_and_digests
```
Comment (clnoll) [inline airbyte-ci/connectors/metadata_service/lib/metadata_service/docker_hub.py:75]: Does the comment on line 73 go with these lines? 

I feel like this could end up confusing us if we need to debug something; naively it looks like this could return a partial response, if we get some results on the first page and then a 404 on the second. I'd prefer to `raise_for_status` on 404s and let this raise an exception that we catch and log in `get_latest_version_on_dockerhub`, and never allow ourselves to return a partial response.

File: airbyte-ci/connectors/metadata_service/lib/tests/test_validators/test_metadata_validators.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
+# Copyright (c) 2024 Airbyte, Inc., all rights reserved.
+
+
+import pytest
+import requests
+import semver
+import yaml
+from metadata_service.models.generated.ConnectorMetadataDefinitionV0 import ConnectorMetadataDefinitionV0
+from metadata_service.validators import metadata_validator
```
Comment (clnoll) [inline airbyte-ci/connectors/metadata_service/lib/tests/test_validators/test_metadata_validators.py:9]: I'd love to have some unit tests here that don't do a network round trip, and that also handle the case where there's no `latest`.

**Review by clnoll — CHANGES_REQUESTED** [review]
This seems generally correct but I think there are some edge cases that we should handle/test a bit differently (details in comments).

File: airbyte-ci/connectors/metadata_service/lib/metadata_service/docker_hub.py (hunk 2 above)
Comment (alafanechere) [reply airbyte-ci/connectors/metadata_service/lib/metadata_service/docker_hub.py:92]: All connector images published via `airbyte-ci` have `latest` tag:
https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/publish/pipeline.py#L144

File: airbyte-ci/connectors/metadata_service/lib/metadata_service/docker_hub.py (hunk 3 above)
Comment (alafanechere) [reply airbyte-ci/connectors/metadata_service/lib/metadata_service/docker_hub.py:64]: I did not use the API doc for it. As we have tag specific details in [`is_image_on_dockerhub`](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/metadata_service/lib/metadata_service/docker_hub.py#L51) with `/v2/repositories/{image_name}/tags/{tag_name}` I figured that according to rest best practices `/tags` should return a list of tags for this image and it worked.

File: airbyte-ci/connectors/metadata_service/lib/metadata_service/docker_hub.py (hunk 4 above)
Comment (alafanechere) [reply airbyte-ci/connectors/metadata_service/lib/metadata_service/docker_hub.py:75]: > Does the comment on line 73 go with these lines?

Yes.

I added a `raise_for_status` on the for loop exit. This should raise an error when, after retry, a different error status than 404 is returned by the API.
I also printed some messages when we face a 404. (this library has no logger and uses `print` in other context...)

File: airbyte-ci/connectors/metadata_service/lib/metadata_service/docker_hub.py (hunk 2 above)
Comment (alafanechere) [reply airbyte-ci/connectors/metadata_service/lib/metadata_service/docker_hub.py:92]: > So I feel like we shouldn't rely on it here, either. Can we check for the max version instead?

As we also release pre-release image with tag which are not semver it's not possible to accurately determine the max version.
I think it's more correct to:
* Find the digest of the `latest` tag
* Find the version tag which match this digest and consider this version the latest one.

It more accurately matches what's happening in our publish pipeline.

File: airbyte-ci/connectors/metadata_service/lib/tests/test_validators/test_metadata_validators.py (hunk 5 above)
Comment (alafanechere) [reply airbyte-ci/connectors/metadata_service/lib/tests/test_validators/test_metadata_validators.py:9]: > Handle the case where there's not `latest`

This case is handled, if no latest version is found we pass the validation as it means the connector has never been published yet:
https://github.com/airbytehq/airbyte/blob/augustin/07-30-metadata-service_validate_that_the_dockerImageTag_is_not_decremented/airbyte-ci/connectors/metadata_service/lib/metadata_service/validators/metadata_validator.py#L189

Comment (alafanechere) [reply airbyte-ci/connectors/metadata_service/lib/tests/test_validators/test_metadata_validators.py:9]: I added a `test_validation_pass_on_docker_image_no_latest` test. 
I'd prefer to keep the network round trip as if the test fail it might characterize a regression in DockerHub API we'd want to address.

File: airbyte-ci/connectors/metadata_service/lib/metadata_service/docker_hub.py (hunk 3 above)
Comment (clnoll) [reply airbyte-ci/connectors/metadata_service/lib/metadata_service/docker_hub.py:64]: In general I'd prefer to use a documented API. Since we're already using the other one I guess it's fine.

File: airbyte-ci/connectors/metadata_service/lib/metadata_service/docker_hub.py (hunk 2 above)
Comment (clnoll) [reply airbyte-ci/connectors/metadata_service/lib/metadata_service/docker_hub.py:96]: We can filter out pre-releases though, because they have a different naming convention.

Theoretically if `latest` was associated with v1.0 but there was a v1.1 in dockerhub, we'd want to increment past 1.1, right?

File: airbyte-ci/connectors/metadata_service/lib/tests/test_validators/test_metadata_validators.py (hunk 5 above)
Comment (clnoll) [reply airbyte-ci/connectors/metadata_service/lib/tests/test_validators/test_metadata_validators.py:9]: I'm fine having tests with a round trip but think that a separate test where the values for the versions are deterministic is appropriate here.

File: airbyte-ci/connectors/metadata_service/lib/metadata_service/docker_hub.py (hunk 2 above)
Comment (alafanechere) [reply airbyte-ci/connectors/metadata_service/lib/metadata_service/docker_hub.py:96]: Yes but that's not meant to happen as we always make `latest` match the latest version. If it happens it means we have a problem somewhere and this check might help catching it.

Comment (clnoll) [reply airbyte-ci/connectors/metadata_service/lib/metadata_service/docker_hub.py:96]: I think if we want this check to catch a problem like that we should sort by tag and then verify that the digest of the max tag is the same as "latest".

File: airbyte-ci/connectors/metadata_service/lib/tests/test_validators/test_metadata_validators.py (hunk 5 above)
Comment (clnoll) [reply airbyte-ci/connectors/metadata_service/lib/tests/test_validators/test_metadata_validators.py:9]: Great, thanks!

File: airbyte-ci/connectors/metadata_service/lib/metadata_service/docker_hub.py (hunk 2 above)
Comment (alafanechere) [reply airbyte-ci/connectors/metadata_service/lib/metadata_service/docker_hub.py:96]: @clnoll that would be a reasonable check to add, but in POST_UPLOAD_VALIDATORS (running at the end of the publish pipeline), not in PRE_UPLOAD_VALIDATORS.

Comment (clnoll) [reply airbyte-ci/connectors/metadata_service/lib/metadata_service/docker_hub.py:96]: @alafanechere I don't want to block you so I'll go ahead and approve, but I'd like to see us moving away from using or relying on "latest", as it's not best practice due to the ambiguity associated with it. I think this conversation actually demonstrates that in a way, but obviously our usage of it is a bigger problem than just this PR.

**Review by clnoll — APPROVED** [review]
